src/cpp/src/model_runner.hpp

```diff
@@ -129,12 +139,20 @@ class ModelRunner {
             for (size_t seq_id = 0; seq_id < num_running_sequences; ++seq_id) {
                 Sequence::CPtr sequence = running_sequences[seq_id];
 
-                for (size_t token_id = 0, position_id = group_position_id; token_id < num_scheduled_tokens; ++token_id, ++position_id) {
+                for (size_t token_id = 0, position_id = group_position_id; token_id < num_scheduled_tokens; ++token_id, ++position_id, ++gathering_current_index) {
                     // compute token for current sequence
                     input_ids_data[token_id] = position_id < sequence_group->get_prompt_len() ?
                         sequence_group->get_prompt_ids()[position_id] :
                         sequence->get_generated_ids()[position_id - sequence_group->get_prompt_len()];
 
+                    if (matmul_gathering_is_required) {
+                        if (group_position_id + token_id >= prompt_len - 1) {
```

Comment:
I suppose here we need a condition similar to one in sampler itself https://github.com/openvinotoolkit/openvino.genai/blob/master/src/cpp/src/sampler.cpp#L771

and it does not cover speculative decoding case: 
- in case of `requires_sampling` we typically sample 1 token
- or num candidates in case of SD

Reply:
in case `echo = true` (see https://github.com/openvinotoolkit/openvino.genai/blob/master/src/cpp/include/openvino/genai/generation_config.hpp#L44) you need to sample prompt tokens as well